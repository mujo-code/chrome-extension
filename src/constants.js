// Config values
export const MINUTE = 1000 * 60
export const FIVE_MINUTES = MINUTE * 5
export const FOURTY_FIVE_MINUTES = MINUTE * 45
export const HOUR = MINUTE * 60
export const THREE_HOURS = HOUR * 3
export const ALARM_DEFAULT_VALUE = true
export const MAX_BREAKTIMERS = 5
export const CURRENT_SUB_SKU = 'early_subscription'
export const APP_READY_KEY = '__mujo_ready__'

// ALARMS
export const ALARM_KEY = 'MINDFUL_ALARM' // also storage key
export const HEARTBEAT = 'HEARTBEAT'

// DB Keys
export const DATABASE_NAME = 'MUJO'
export const DATABASE_STORE = 'STATS'
export const DATABASE_VERSION = 1
export const LAST_ACTIVITY_TABLE = 'LAST_ACTIVITY'

// Storage keys
export const TOP_SITES_KEY = 'TOP_SITES'
export const TOP_SITES_USAGE_KEY = 'TOP_SITES_USAGE'
export const PAGE_VIEWS_KEY = 'PAGE_VIEWS'
export const SHOW_TOP_SITES_KEY = 'SHOW_TOP_SITES'
export const SITE_TIME_KEY = 'SITE_TIME'
export const BREAK_TIMERS_KEY = 'BREAK_TIMERS'
export const LAST_ACTIVE_KEY = 'LAST_ACTIVE_KEY'
export const ACTIVITY_NUMBER_KEY = 'ACTIVITY_NUMBER_KEY'

// Actions
export const NEW_TAB_CONNECTION = 'New Connection'
export const PAGE_VIEWING_TIME = 'Page Viewing Time'
export const DEEP_LINK_NEWTAB = 'Deep Link New Tab'
export const RESET_USAGE = 'Reset Usage Stats'
export const TRACK = 'Track'
export const ADD_BROADCAST_TAB = 'Add Broadcast Tab'
export const REMOVE_BROADCAST_TAB = 'Remove Broadcast Tab'
export const VALUE_CHANGED = 'Value changed'

// Getter, Setters for storage
export const GET_STORAGE = 'GET_STORAGE'
export const SET_STORAGE = 'SET_STORAGE'

// Feature Flags
export const SCREEN_TIME_FEATURE = true
export const BREAK_TIMER_FEATURE = true
export const SUBSCRIBE_FEATURE = false

// Upsell modals
export const MAX_BREAKTIMER_MODAL = 'breakTimerMax'
export const SUB_DETAILS_MODAL = 'subscriptionDetails'
export const SUB_SUCCESS_MODAL = 'subSuccess'

// OPTIONAL permissions requested
export const SCREEN_TIME_PERMISSIONS = {
  permissions: ['activeTab'],
  origins: ['https://*/*', 'http://*/*'],
}
