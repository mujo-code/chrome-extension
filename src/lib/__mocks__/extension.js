/*
  Extension Lib Mock
*/
import model from '../../model'

export const getStorageFn = async key => model[key].defaultValue
export const setStorageFn = async (key, value) => value

export const message = jest.fn().mockResolvedValue({})
export const getStorage = jest.fn().mockImplementation(getStorageFn)
export const setStorage = jest.fn().mockImplementation(setStorageFn)
export const onMessage = jest.fn()

export const alarms = {
  get: jest.fn(),
  create: jest.fn(),
  clear: jest.fn(),
}
export const notifications = {}
export const tabs = {}
export const webNavigation = {}
export const runtime = {}
export const topSites = { get: jest.fn() }
