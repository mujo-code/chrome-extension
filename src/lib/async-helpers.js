export const wait = (amount = 0) =>
  new Promise(resolve => {
    setTimeout(resolve, amount)
  })

export const defer = async (fn, amount, ...args) => {
  await wait(amount)
  return fn(...args)
}

export const composePromises = (...fns) => value =>
  fns.reduceRight(
    (promise, fn) => promise.then(resolve => fn(resolve)),
    Promise.resolve(value)
  )

export const pipePromises = (...fns) => composePromises(fns.reverse())
