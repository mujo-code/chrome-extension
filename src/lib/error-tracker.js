import * as Sentry from '@sentry/browser'

const { SENTRY_DSN } = process.env

export const createTracker = ({ dsn, environment }) => {
  const hasDSN = !!dsn
  if (hasDSN) {
    Sentry.init({ dsn, environment })
  }

  return {
    exception(error) {
      if (hasDSN) return null
      return Sentry.captureException(error)
    },
    async exceptionWithInfo(error, errorInfo) {
      if (hasDSN) return null
      return new Promise(resolve => {
        Sentry.withScope(scope => {
          scope.setExtras(errorInfo)
          return resolve(Sentry.captureException(error))
        })
      })
    },
  }
}

export const tracker = createTracker({ dsn: SENTRY_DSN })
