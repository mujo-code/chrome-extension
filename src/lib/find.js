import { curry, first } from './functional'

export const find = curry((query, arr) => {
  const keys = Object.keys(query)
  return first(
    arr.filter(item => keys.every(key => item[key] === query[key]))
  )
})
