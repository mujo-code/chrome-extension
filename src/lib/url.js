import punycode from 'punycode'

export const parse = url => new URL(url)
export const shortURL = url => punycode.toUnicode(parse(url).hostname)
export const origin = url => parse(url).origin

export const queryParams = url => {
  const { search = '?' } = parse(url)
  return search
    .slice(1)
    .split(/&/g)
    .map(keyvalue => {
      const [key, value] = keyvalue.split('=')
      return { [key]: decodeURIComponent(value) }
    })
    .reduce((accum, values) => {
      Object.assign(accum, values || {})
      return accum
    }, {})
}
