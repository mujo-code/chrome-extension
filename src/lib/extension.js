import { GET_STORAGE, SET_STORAGE } from '../constants'
import { promisifyObject } from './promisify'
/*
  Extension Lib
  -----
  Essentially a way to abstract away chrome bits
  to eventually allow for bits like firefox without large refactors
*/

export const message = async (event, data) =>
  new Promise((resolve, reject) => {
    const payload = Object.assign({ event }, data || {})
    chrome.runtime.sendMessage(payload, (response = {}, ...args) => {
      console.log([response], args)
      if (response.error) {
        reject(response.error)
        return
      }
      resolve(response)
    })
  })

export const getStorage = async key => {
  const response = await message(GET_STORAGE, { key })
  return response.value
}

export const setStorage = async (key, value) => {
  const response = await message(SET_STORAGE, { key, value })
  return response.value
}

export const onMessage = (...args) => {
  chrome.runtime.onMessage.addListener(...args)
}

export const permissions = promisifyObject(chrome.permissions)

export const {
  alarms,
  notifications,
  tabs,
  webNavigation,
  runtime,
  topSites,
} = chrome
