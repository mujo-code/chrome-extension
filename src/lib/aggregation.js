import { shortURL } from './url'
import { set } from './util'

const combineUrlTimes = data => {
  const keys = Object.keys(data)
  return keys.reduce((accum, key) => {
    const shortKey = shortURL(key)
    set(accum, shortKey, data[key] + (accum[shortKey] || 0))
    return accum
  }, {})
}

const getTotalTime = data => {
  const iterable = Object.keys(data)
  return iterable
    .map(key => data[key])
    .reduce((total, value) => {
      /* eslint-disable-next-line no-param-reassign */
      total += value
      return total
    }, 0)
}

export const siteTimeToChartData = rawData => {
  const data = combineUrlTimes(rawData)
  const totals = getTotalTime(data)
  return Object.keys(data).map(label => {
    const percent = data[label] / totals
    return {
      label,
      percent,
      originalData: { [label]: data[label] },
    }
  })
}
