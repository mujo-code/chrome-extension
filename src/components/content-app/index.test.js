import React from 'react'
import { act, create } from 'react-test-renderer'
import { BREAK_TIMERS_KEY, SITE_TIME_KEY } from '../../constants'
import { getStorage } from '../../lib/extension'
import { origin } from '../../lib/url'
import ContentApp from '.'

const CONTAINER_ID = 'mujo-extension'
const URL = origin(window.location.href)

beforeEach(() => {
  const el = document.createElement('div')
  el.id = CONTAINER_ID
  document.body.appendChild(el)
})

afterEach(() => {
  const el = document.getElementById(CONTAINER_ID)
  if (el) {
    el.remove()
  }
})

test('ContentApp should match snapshot', () => {
  const tree = create(<ContentApp />).toJSON()
  expect(tree).toMatchSnapshot()
})
describe('Async ContentApp', () => {
  beforeAll(() => {
    getStorage.mockReset()
  })
  afterAll(() => {
    getStorage.mockReset()
  })
  test('ContentApp should match snapshot', async () => {
    const timer = { enabled: true, time: 1 }
    const results = {
      [BREAK_TIMERS_KEY]: { [URL]: timer },
      [SITE_TIME_KEY]: { [URL]: 2 },
    }
    const mockResults = async key => results[key]
    const storagePromise = new Promise(resolve => {
      getStorage
        .mockImplementationOnce(mockResults)
        .mockImplementationOnce(mockResults)
        .mockImplementationOnce(mockResults)
        .mockImplementationOnce(key => {
          resolve()
          return mockResults(key)
        })
    })

    let container

    act(() => {
      container = create(<ContentApp />)
    })
    await storagePromise
    expect(container.toJSON()).toMatchSnapshot()
  })
})
