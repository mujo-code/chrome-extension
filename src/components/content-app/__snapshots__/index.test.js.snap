// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async ContentApp ContentApp should match snapshot 1`] = `
<div
  className="css-1yv2jt6 css-16qvk6d css-83d5g6"
  data-mujo={true}
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="mujo-modal css-1m3lehq"
  shouldCloseOnOverlayClick={true}
>
  <svg
    className="css-1qcxu9b css-1ouy01g"
    height={20}
    onClick={[Function]}
    width={19}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="css-lv59v1"
      d="
        M9.778 6.864
        l5.754-6.1
        a2 2 0 1 1 2.91 2.746
        l-5.914 6.268 5.913 6.269
        a2 2 0 1 1-2.91 2.744
        l-5.753-6.098-5.753 6.098
        a2 2 0 1 1-2.91-2.744
        l5.914-6.269L1.115 3.51
        A2 2 0 1 1 4.025.765l5.753 6.099z
        "
      fillRule="nonzero"
    />
  </svg>
  <div
    className="css-1klwfzq css-tk92wv"
    onClick={[Function]}
  >
    <svg
      className="css-1f61c8m css-8a9gku"
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        className="css-c901fj css-6xifmd"
      />
      <circle
        className="css-1xu6mku css-1yq83d0"
      />
      <circle
        className="css-lv59v1 css-1yq83d0"
      />
      <g
        className="css-1tb56z6"
      >
        <text
          className="css-1d0qo6a css-ou8fjb css-bsai6x css-c901fj css-kezvte css-f02zlw css-16uyi9m css-mhk3h4"
          textAnchor="middle"
          x={120}
          y={128}
        >
          Take a break
        </text>
      </g>
    </svg>
  </div>
  <div
    className="css-vqh3cg"
  >
    <p
      className="css-qpwegv css-6xdpdk css-1ypmooz css-h89647 css-nxjalj css-2zzudi css-mhk3h4"
    >
      The site 
      localhost
       was actively viewed for
       
      <span
        className="css-hukeue"
      >
        2 milliseconds
      </span>
       
      since your last break. Right now is a good time to take a break
    </p>
    <button
      className="css-1vu1all css-1lkqdt0 css-1udh8mm css-h0biv2 css-5uyuv6 css-1k8tu3j css-1okiobr css-yxn9bq css-1pev02g css-7ep5t css-b9abf0 css-1xg3sco css-1dnwvu3 css-13azwyo css-1kvvogt css-ou8fjb"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Edit site timer
    </button>
  </div>
</div>
`;

exports[`ContentApp should match snapshot 1`] = `null`;
